File: 100_same_tree/100.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        std::queue<TreeNode*> p_queue;
        std::queue<TreeNode*> q_queue;
        p_queue.push(p);
        q_queue.push(q);
        
        while (!p_queue.empty() && !q_queue.empty()) {
            TreeNode* p = p_queue.front();
            p_queue.pop();
            
            TreeNode* q = q_queue.front();
            q_queue.pop();
            
            if (
                (p == nullptr && q != nullptr)
                || (p != nullptr && q == nullptr)
            ) {
                return false;
            } else if (p != nullptr) {
                if (p->val != q->val) {
                    return false;
                }
                p_queue.push(p->left);
                p_queue.push(p->right);
                q_queue.push(q->left);
                q_queue.push(q->right);
            }
        }
        return true;
    }
};

File: 100_same_tree/100.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        p_queue = [p]
        q_queue = [q]
        
        while p_queue and q_queue:
            p = p_queue.pop(0)
            q = q_queue.pop(0)
            
            if bool(p) != bool(q):
                return False
            elif p is not None:
                if p.val != q.val:
                    return False
                
                p_queue.append(p.left)
                p_queue.append(p.right)
                q_queue.append(q.left)
                q_queue.append(q.right)
        return True
